pkg/components/util/delete.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2020 The Lokomotive Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"fmt"
+	"strings"
+
+	"helm.sh/helm/v3/pkg/action"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kinvolk/lokomotive/pkg/components"
+	"github.com/kinvolk/lokomotive/pkg/k8sutil"
+)
+
+// DeleteComponent deletes a component
+func DeleteComponent(name string, c components.Component, kubeconfig string) error {
+	cs, err := k8sutil.NewClientset(kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	ns := c.Metadata().Namespace
+	if ns == "" {
+		return fmt.Errorf("component %s namespace is empty", name)
+	}
+
+	cfg, err := HelmActionConfig(ns, kubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed preparing helm client: %w", err)
+	}
+
+	var errs []string
+
+	uninstall := action.NewUninstall(cfg)
+
+	resp, err := uninstall.Run(name)
+	if err != nil {
+		errs = append(errs, err.Error())
+	} else {
+		fmt.Println(resp.Info)
+	}
+
+	// Ensure we delete the manually created namespace which was not created by helm
+	if err = cs.CoreV1().Namespaces().Delete(ns, &metav1.DeleteOptions{}); err != nil {
+		errs = append(errs, err.Error())
+	}
+
+	return AggregateErrors(errs)
+}
+
+// AggregateErrors conflates multiple errors stored in a string form and returns a single error.
+func AggregateErrors(errs []string) error {
```

Comment:
Why not use `hcl.Diagnostics` instead?

Reply:
I think hcl.Diagnostics is not very helpful to the end user because it does not list all the errors and oly gives output as follows:

```
FATA[0016] <nil>: uninstall: Release not loaded: openebs-operator: release: not found; , and 4 other diagnostic(s)  args="[]" command="lokoctl component delete"
```

This is not very helpful. As a user I want to know all the errors that have occurred.

Reply:
@surajssd you need to iterate over it and print each one in desired format to make it usable for the user :wink:

Reply:
Then the API is not very useful. Because this function `AggregateErrors` clubs all the errors one on each line and the output looks like following:

```
$ lokoctl component delete --delete-namespace
Do you want to delete components: "openebs-operator,openebs-storage-class,prometheus-operator,contour,metallb,cert-manager,httpbin"? [type "yes" to continue]:
 yes
Deleting component 'openebs-operator'...
Deleting component 'openebs-storage-class'...
Deleting component 'prometheus-operator'...
Deleting component 'contour'...
Deleting component 'metallb'...
Deleting component 'cert-manager'...
Deleting component 'httpbin'...

FATA[0008] uninstall: Release not loaded: openebs-operator: release: not found
uninstall: Release not loaded: openebs-storage-class: release: not found
uninstall: Release not loaded: prometheus-operator: release: not found
namespaces "monitoring" not found
uninstall: Release not loaded: contour: release: not found
namespaces "projectcontour" not found
uninstall: Release not loaded: metallb: release: not found
namespaces "metallb-system" not found
uninstall: Release not loaded: cert-manager: release: not found
namespaces "cert-manager" not found
uninstall: Release not loaded: httpbin: release: not found
namespaces "httpbin" not found  args="[]" command="lokoctl component delete"
```

Reply:
¯\_(ツ)_/¯

Reply:
I think we don't need this function - see https://github.com/kinvolk/lokomotive/pull/268#discussion_r408074765.